contracts/BEP20RewardApeV5.sol

```diff
@@ -0,0 +1,288 @@
+        if (_to <= bonusEndBlock) {
+            return _to - _from;
+        } else if (_from >= bonusEndBlock) {
+            return 0;
+        } else {
+            return bonusEndBlock - _from;
+        }
+    }
+
+    /// @param  _bonusEndBlock The block when rewards will end
+    function setBonusEndBlock(uint256 _bonusEndBlock) external onlyOwner {
+        require(_bonusEndBlock > block.number, 'new bonus end block must be greater than current');
+        bonusEndBlock = _bonusEndBlock;
+        emit LogUpdatePool(bonusEndBlock, rewardPerBlock);
+    }
+
+    // View function to see pending Reward on frontend.
+    function pendingReward(address _user) external view returns (uint256) {
+        UserInfo storage user = userInfo[_user];
+        uint256 accRewardTokenPerShare = poolInfo.accRewardTokenPerShare;
+        if (block.number > poolInfo.lastRewardBlock && totalStaked != 0) {
+            uint256 multiplier = getMultiplier(poolInfo.lastRewardBlock, block.number);
+            uint256 tokenReward = multiplier * rewardPerBlock * poolInfo.allocPoint / totalAllocPoint;
+            accRewardTokenPerShare = accRewardTokenPerShare + (tokenReward * 1e30 / totalStaked);
+        }
+        return user.amount * accRewardTokenPerShare / 1e30 - user.rewardDebt;
+    }
+
+    // Update reward variables of the given pool to be up-to-date.
+    function updatePool() public {
+        if (block.number <= poolInfo.lastRewardBlock) {
+            return;
+        }
+        if (totalStaked == 0) {
+            poolInfo.lastRewardBlock = block.number;
+            return;
+        }
+        uint256 multiplier = getMultiplier(poolInfo.lastRewardBlock, block.number);
+        uint256 tokenReward = multiplier * rewardPerBlock * poolInfo.allocPoint / totalAllocPoint;
+        totalRewardsAllocated += tokenReward;
+        poolInfo.accRewardTokenPerShare = poolInfo.accRewardTokenPerShare + (tokenReward * 1e30 / totalStaked);
+        poolInfo.lastRewardBlock = block.number;
+    }
+
+
+    /// Deposit staking token into the contract to earn rewards.
+    /// @dev Since this contract needs to be supplied with rewards we are
+    ///  sending the balance of the contract if the pending rewards are higher
+    /// @param _amount The amount of staking tokens to deposit
+    function deposit(uint256 _amount, address _user) external nonReentrant {
+        require(_user != address(0), "Can't deposit for null address");
+
+        UserInfo storage user = userInfo[_user];
+        updatePool();
+        if (user.amount > 0) {
+            uint256 pending = user.amount * poolInfo.accRewardTokenPerShare / 1e30 - user.rewardDebt;
+            if(pending > 0) {
+                // If rewardBalance is low then revert to avoid losing the user's rewards
+                require(rewardBalance() >= pending, "insufficient reward balance");
+                safeTransferRewardInternal(_user, pending);
```

Comment:
**MEDIUM** Vulnerability

Most account approve the `MAX_UINT` for contract allowances. This function allows ANY `external` user to deposit funds from other users accounts.

`depositTo` would allow an external contract to deposit tokens for another user, so `depositTo` should transfer tokens from `msg.sender`.
- [ ] `safeTransferRewardInternal(msg.sender, pending);`

Reply:
safeTransferRewardInternal are transfer of rewards to owner. with msg.sender here you can steal rewards of others

the tokens to deposit are still coming from sender
`poolInfo.lpToken.safeTransferFrom(address(msg.sender), address(this), _amount);`

Reply:
Ahh nice, thanks!